### ISSUE TITLE
Unable to reproduce reported LIBERO-Spatial/Object success rates with released checkpoints

### ISSUE THREAD

**Issue by ShawRong** [H-EmbodVis/TurboVLA#4]
Hi, and thanks a lot for releasing the code and checkpoints. The setup was easy to follow, and everything loaded cleanly on the first try.

We are trying to reproduce the reported LIBERO results, but are seeing substantially lower success rates than those in the paper. We suspect there may be an evaluation detail we are missing and would really appreciate any pointers.

## Results

Using the released checkpoints (`spatial.pth` / `object.pth`), with all 10 tasks per suite and 50 trials per task:

| Suite | Ours | Paper |
|---|---:|---:|
| LIBERO-Spatial | 77.2% (386/500) | 99.2% |
| LIBERO-Object | 69.2% (346/500) | 99.8% |

Per-task success rates:

```text
spatial: 92  86  78  96   4  64  98  98  98  58   (mean 77.2)
object:  70  92  94  50  56  88  68  12  78  84   (mean 69.2)
````

Most tasks are in a reasonable range, but a few collapse—for example, Spatial task 4 at 4% and Object task 7 at 12%—which pulls down the suite average substantially.

## Evaluation command

We use the command from the README:

```bash
python experiments/libero/evaluate.py \
  --ckpt_path pretrained/TurboVLA/checkpoints/libero/spatial.pth \
  --dinov3_path <local dinov3-vitb16> \
  --bert_path <local bert-base-uncased> \
  --stats_path experiments/libero/configs/libero_all4_stats.json \
  --stats_key libero_all4_no_noops \
  --task_suite_name libero_spatial \
  --num_trials_per_task 50 \
  --chunk_size 12 \
  --num_open_loop_steps 12 \
  --seed 7 \
  --precision bf16
```

We shard evaluation across GPUs using `--task_ids`.

Since `eval_libero()` calls `set_seed_everywhere(cfg.seed)` at the beginning of every task, we believe this should be equivalent to evaluating the full suite in a single process.

### Checkpoints

* Verified all four `.pth` files against the SHA-256 hashes from the Hub.
* `load_state_dict(..., strict=True)` succeeds.
* 672 tensors loaded with no missing or unexpected keys.

### Normalization statistics

* `experiments/libero/configs/libero_all4_stats.json` is byte-identical to the copy in the Hugging Face release.
* Confirmed `suite_policy.py` is using these statistics rather than the module-level constants in `policy.py`.

### Text processing

* Per-instruction padding is taken from the checkpoint's embedded `model_config`.
* All 40 LIBERO instructions are present in the table.
* We do not see any correlation between padding length and success rate.

### Renderer

We tested both:

```text
--mujoco_gl osmesa
--mujoco_gl egl
```

The two renderers produce visibly different frames (~81% of pixels differ for the same scene), but their rollout success rates are the same within noise.

We also compared our rendered observations against LIBERO demonstration frames. Brightness and per-channel statistics are very similar:

```text
demo mean: 115.98
ours mean: 116.68
```

So the visual input distribution appears reasonable.

### Open-loop horizon

We tested:

```text
--num_open_loop_steps 12
--num_open_loop_steps 4
--num_open_loop_steps 1
```

None improved the results significantly.

### Precision

Using `fp32` instead of `bf16` improves some of the worst-performing tasks slightly, but does not come close to closing the gap.

### Model / preprocessing sanity check

As an additional check, we ran a teacher-forcing-style evaluation: feeding observations from real LIBERO demonstrations into the released policy and comparing the predicted actions with demonstration actions.

We observe correlations of approximately `0.85–0.99` across action dimensions, and the predicted action magnitudes closely match the demonstrations.

This makes us think the following parts of the pipeline are probably functioning correctly:

* observation preprocessing
* normalization / denormalization
* text conditioning
* checkpoint loading
* model forward pass

## Observed failure mode

In rollout videos, a common failure pattern is:

1. the arm reaches the correct object,
2. successfully grasps it,
3. then hovers near the grasp location,
4. and does not transport the object to the target before the episode reaches the step limit.

This is especially common on the tasks with very low success rates.

## Questions

1. **Is there anything beyond the README evaluation command required to reproduce the reported 99.2% / 99.8% success rates?**

2. **Which `robosuite` and `mujoco` versions were used for the reported evaluation?**

   We are currently using:

   ```text
   robosuite==1.4.1
   mujoco==3.1.6
   ```

3. **How should we interpret the reported 31.2 ms / 32 Hz latency relative to `--num_open_loop_steps 12`?**

   The 31.2 ms figure sounds like per-control-step policy inference, while the provided LIBERO evaluation command executes 12 predicted actions open-loop.

   Was the reported latency measured with:

   * one policy inference every environment step, or
   * one policy inference every 12 environment steps?

4. **Would it be possible to share the per-task success rates for one of the LIBERO suites?**

   That would help us determine whether we have:

   * a global degradation across all tasks, or
   * a small number of tasks whose behavior differs significantly from the reported evaluation.

## Environment

The main environment difference we could not eliminate is the PyTorch/CUDA stack.

We are running on an RTX 5090 (`sm_120`), which requires CUDA 12.8+, so our environment uses:

```text
torch==2.7.1+cu128
transformers==4.57.6
```

instead of the reference environment:

```text
torch==2.3.1+cu121
transformers==4.56
```

The teacher-forcing sanity check above was partly intended to test whether this stack difference was causing a major model discrepancy, but we did not observe evidence of that.

LIBERO is currently at upstream master:

```text
8f1084e
```

## Minor note: `normalize_binary_gripper`

Unrelated to the reproduction gap, but we noticed that in `turbovla/evaluation/suite_policy.py` the two branches below are currently identical:

```python
if self.normalize_binary_gripper:
    gripper_source = float(action_norm[6])
else:
    gripper_source = float(action_norm[6])
```

So `normalize_binary_gripper` does not currently affect this code path.

In our runs the flag resolves to `False`, so this does not appear to explain our results. Just flagging it in case the two branches were intended to behave differently.

Thanks again for the release and for any guidance!

**Comment by IMMORTAALLL** [comment]
Thanks for the detailed report. We identified a compatibility issue.

The released TurboVLA code was evaluated with Transformers 4.56.0, while your setup uses 4.57.6. The newer version changed the behavior of DINOv3's `outputs.hidden_states[-1]`, which TurboVLA uses as its visual feature. Therefore, strict checkpoint loading can succeed while the predicted actions differ.

Please try:

    pip install --force-reinstall "transformers==4.56.0"

We are preparing a patch to read `last_hidden_state` explicitly and to pin/check the supported Transformers version. Thank you for helping us identify this issue.

**Comment by ShawRong** [comment]
Copy that, thanks for ur patience.
